File: DS/Linkedlist/doubly_linked_list/CPP/doubly_linked_list_implementation.cpp
#include <iostream>
using namespace std;

//node structure 
struct node{
    int data;
    node* prev;
    node* next;

    //node constructor
    node(int num){
        data = num;
        prev = NULL;
        next = NULL;
    }
};

node* head;          //head pointer
int nodeCount=0;     //node counter

//insert node at the end of doublylinked list
void insertAtEnd(int data){
    node* temp = new node(data);    //create node
    nodeCount++;
    if(head == NULL){
        head = temp;
        return;
    }

    node* temp1 = head;
    while (temp1->next!=NULL)   //going at last node
    {
        temp1 = temp1->next;    
    }
    
    // linking new node with list
    temp1->next = temp;         
    temp->prev = temp1;
}

//insert node at the front of doublylinked list
void insertAtHead(int data){
    node* temp = new node(data);    //create node
    nodeCount++;
    if(head==NULL){
        head = temp;
        return;
    }
    // linking new node with list
    head->prev = temp;
    temp->next = head;
    head = temp;                //changing head pointer
}

void insert(int index,int data){
    //check index value
    //if index is invalid then exit
    if(index==1||index>nodeCount){
        cout<<"Invalid index";
        return;
    }

    // else create node
    node* temp = new node(data);
    index -=2;

    node* temp1 = head;
    while (index--)              //going at (index-1) node
    {
        temp1 = temp1->next;
    }

    // linking new node with list
    temp->next = temp1->next;
    temp->prev = temp1;
    temp1->next = temp;
    nodeCount++;
}

//print list elements
void print(){
    node* temp = head;
    while (temp!=NULL)
    {
        cout<<temp->data<<" ";
        temp = temp->next;
    }
    cout<<"\n";
}

int main(){
    head = NULL;
    insertAtEnd(2);
    insertAtHead(1);
    insertAtEnd(3);
    insertAtEnd(4);
    insert(2,6);
    cout<<"List: ";
    print();
}

// output
// List: 1 6 2 3 4

File: DS/Linkedlist/doubly_linked_list/JAVA/doubly_linked_list_implementation.java
//create class DoublyLinkedList
class DoublyLinkedList{
        //node class to create node
        static class node{
            int data;
            node next;
            node prev;

            //Constructor to create a new node
            node(int num){
                data = num;
                prev = null;
                next = null;
            }
        }

        node head;              //head pointer
        int nodeCount = 0;      //nodeCount counts number of nodes

        //insert node at the front of doublylinked list
        public void insertAtHead(int data){
            node temp = new node(data);     //create new node

            if(head!=null){
                temp.next = head;
                head.prev = temp;
            }
            head = temp;
            nodeCount++;
        }

        //insert node at the end of doublylinked list
        public void insertAtEnd(int data){
            node temp = new node(data);     //create new node
            if(head==null){
                head = temp;
            }
            else{
                node temp1 = head;

                //going at last node
                while (temp1.next!=null){
                    temp1 = temp1.next;
                }
                //changing links to add new node at end
                temp1.next = temp;
                temp.prev = temp1;
            }
            nodeCount++;
        }

        //insert node at given index
        public void insert(int index, int data){
            if(index>nodeCount){
                System.out.println("Invalid index");
                return;
            }
            node temp = new node(data);
            if (index==1){
                head.prev = temp;
                temp.next = head;
                head = temp;
                return;
            }

            index -= 2;
            node temp1 = head;

            //going at (index-1) node
            while ((index--)>0){
                temp1 = temp1.next;
            }
            //changing links to add new node at given index
            temp.next = temp1.next;
            temp.prev = temp1;
            temp1.next = temp;
            nodeCount++;
        }

        //print linked list
        public void print(){
            node temp = head;
            while (temp!=null){
                System.out.print(temp.data+" ");
                temp = temp.next;
            }
            System.out.print("\n");
        }
}

public class Main {

    public static void main(String[] args) {
        DoublyLinkedList dll = new DoublyLinkedList();  //create empty list
        dll.head = null;

        //insert elements
        dll.insertAtHead(1);
        dll.insertAtEnd(3);
        dll.insertAtEnd(4);
        dll.insert(2,2);
        System.out.print("list: ");
        dll.print();
    }
}

//output:
//list: 1 2 3 4
